packages/tools/src/utilities/cine/playClip.ts

```diff
@@ -262,4 +296,143 @@ function _stopClipWithData(playClipData) {
   }
 }
 
+function _getVolumeFromViewport(viewport): Types.IImageVolume {
+  const actorEntry = viewport.getDefaultActor();
+
+  if (!actorEntry) {
+    console.warn('No actor found');
+  }
+
+  const volumeId = actorEntry.uid;
+  return cache.getVolume(volumeId);
+}
+
+function _createStackViewportCinePlayContext(
+  viewport: StackViewport
+): CINETypes.CinePlayContext {
+  const imageIds = viewport.getImageIds();
+
+  return {
+    get numScrollSteps(): number {
+      return imageIds.length;
+    },
+    get currentStepIndex(): number {
+      return viewport.getTargetImageIdIndex();
+    },
+    get frameTimeVectorEnabled(): boolean {
+      // It is always in acquired orientation
+      return true;
+    },
+    scroll(delta: number): void {
+      scroll(viewport, { delta, debounceLoading: debounced });
+    },
+  };
+}
+
+function _createVolumeViewportCinePlayContext(
+  viewport: VolumeViewport,
+  volume: Types.IImageVolume
+): CINETypes.CinePlayContext {
+  const actorEntry = viewport.getDefaultActor();
+
+  if (!actorEntry) {
+    console.warn('No actor found');
+  }
+
+  const volumeId = actorEntry.uid;
+  const cachedScrollInfo = {
+    viewPlaneNormal: vec3.create(),
+    scrollInfo: null,
+  };
+
+  const getScrollInfo = () => {
+    const camera = viewport.getCamera();
+    const updateCache =
+      !cachedScrollInfo.scrollInfo ||
+      !vec3.equals(camera.viewPlaneNormal, cachedScrollInfo.viewPlaneNormal);
+
+    // Number of steps would change only after rotating the volume so it
+    // caches the result and recomputes only when necessary. Until it is
+    // rotated the current frame is updated locally
+    if (updateCache) {
+      const scrollInfo = csUtils.getVolumeViewportScrollInfo(
+        viewport,
+        volumeId
+      );
+
+      cachedScrollInfo.viewPlaneNormal = camera.viewPlaneNormal;
+      cachedScrollInfo.scrollInfo = scrollInfo;
+    }
+
+    return cachedScrollInfo.scrollInfo;
+  };
+
+  return {
+    get numScrollSteps(): number {
+      return getScrollInfo().numScrollSteps;
```

Comment:
should we call `getScrollInfo` couple of times here and below? or we can just call that it once and re-use it ?

Reply:
`numScrollSteps` and `currentStepIndex` are accessed on every frame. I thought about creating a method to return both and reduce the number of calls to `getScrollInfo` but after caching the data that solved the problem because the only thing it does is compare the `viewPlanesNormal`s which are a very cheap operation. If you want I can change but that will not make any different in terms of performance (nanoseconds range).